datafusion/sql/src/expr/subquery.rs

```diff
@@ -30,13 +30,15 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         input_schema: &DFSchema,
         planner_context: &mut PlannerContext,
     ) -> Result<Expr> {
+        let old_outer_query_schema = planner_context.outer_query_schema.clone();
+        planner_context.outer_query_schema = Some(input_schema.clone());
+        let sub_plan = self.query_to_plan(subquery, planner_context)?;
+        let outer_ref_columns = sub_plan.all_out_ref_exprs();
+        planner_context.outer_query_schema = old_outer_query_schema;
```

Comment:
Can we support the following subquery in the future? 
```sql
# three level
# the innermost(`t3`) reference outermost(`t1`) column
SELECT *
FROM t1
WHERE EXISTS (
	SELECT *
	FROM t2
	WHERE t1_id = t2_id
		AND EXISTS (
			SELECT *
			FROM t3
			WHERE t3_id > t1_id
		)
);
```

I just think the optimizer rule you plan to add(https://github.com/apache/arrow-datafusion/issues/5492) can also optimize it.

Reply:
I'm afraid I will not support this in recent.   I think the method mentioned in the "Unnesting Arbitrary Queries" paper can cover this, but there is additional complexity need to deal with if there are some ambiguous to the out ref columns,  because different level of out plans' schema might have ambiguous columns.